File: android/src/main/java/com/memotalk/sdk/FileManager.kt
package com.memotalk.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

class FileManager(private val bleClient: BleClient, private val cacheDir: File) {

    suspend fun downloadRecording(device: Device, recording: Recording, progressCallback: (progress: Int) -> Unit = {}): Result<File, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x84: Get local audio file data
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x84 // Command

            bleClient.sendCommand(command)

            // File transfer logic will be implemented
            // This should handle:
            // 1. Request file data via BLE
            // 2. Receive data chunks via FFE0 service (0x23 command)
            // 3. Handle retransmission (0xA3 command)
            // 4. Save to local file

            val outputFile = File(cacheDir, "${recording.fileId}.opus")
            progressCallback(100)
            Result.success(outputFile)
        } catch (e: Exception) {
            Result.failure(PlaudError(5001, "Download recording failed: ${e.message}"))
        }
    }

    suspend fun deleteRecording(device: Device, recording: Recording): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x92: FTP file delete
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x92 // Command

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(5002, "Delete recording failed: ${e.message}"))
        }
    }

    suspend fun getStorageInfo(device: Device): Result<Pair<Long, Long>, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x0F: Get storage space
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x0F // Command

            bleClient.sendCommand(command)
            // Response parsing logic will be implemented
            Result.success(Pair(1000000L, 10000000L)) // (free, total)
        } catch (e: Exception) {
            Result.failure(PlaudError(5003, "Get storage info failed: ${e.message}"))
        }
    }

    suspend fun uploadRecordingToCloud(recording: Recording, filePath: String): Result<String, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Upload logic to cloud service (e.g., Volcano Engine)
            // This will be implemented after cloud API details are provided
            Result.success("upload-id-123")
        } catch (e: Exception) {
            Result.failure(PlaudError(5004, "Upload to cloud failed: ${e.message}"))
        }
    }
}

File: ios/FileManager.swift
import Foundation

public class FileManager {

    private let bleClient: BleClient
    private let cacheDirectory: URL

    public init(bleClient: BleClient, cacheDirectory: URL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]) {
        self.bleClient = bleClient
        self.cacheDirectory = cacheDirectory
    }

    public func downloadRecording(device: Device, recording: Recording, progressCallback: @escaping (Int) -> Void = { _ in }) async -> Result<URL, PlaudError> {
        do {
            // Command 0x84: Get local audio file data
            let command = Data([0x5A, 0x03, 0x84])
            bleClient.sendCommand(command: command)

            // File transfer logic will be implemented
            // This should handle:
            // 1. Request file data via BLE
            // 2. Receive data chunks via FFE0 service (0x23 command)
            // 3. Handle retransmission (0xA3 command)
            // 4. Save to local file

            let outputURL = cacheDirectory.appendingPathComponent("\(recording.fileId).opus")
            progressCallback(100)
            return .success(outputURL)
        } catch {
            return .failure(PlaudError(code: 5001, message: "Download recording failed: \(error.localizedDescription)"))
        }
    }

    public func deleteRecording(device: Device, recording: Recording) async -> Result<Void, PlaudError> {
        do {
            // Command 0x92: FTP file delete
            let command = Data([0x5A, 0x03, 0x92])
            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 5002, message: "Delete recording failed: \(error.localizedDescription)"))
        }
    }

    public func getStorageInfo(device: Device) async -> Result<(free: Int64, total: Int64), PlaudError> {
        do {
            // Command 0x0F: Get storage space
            let command = Data([0x5A, 0x03, 0x0F])
            bleClient.sendCommand(command: command)
            // Response parsing logic will be implemented
            return .success((free: 1000000, total: 10000000))
        } catch {
            return .failure(PlaudError(code: 5003, message: "Get storage info failed: \(error.localizedDescription)"))
        }
    }

    public func uploadRecordingToCloud(recording: Recording, filePath: String) async -> Result<String, PlaudError> {
        do {
            // Upload logic to cloud service (e.g., Volcano Engine)
            // This will be implemented after cloud API details are provided
            return .success("upload-id-123")
        } catch {
            return .failure(PlaudError(code: 5004, message: "Upload to cloud failed: \(error.localizedDescription)"))
        }
    }
}
